src/include/duckdb/function/table_function.hpp

```diff
@@ -329,15 +331,57 @@ enum class TableFunctionInitialization { INITIALIZE_ON_EXECUTE, INITIALIZE_ON_SC
 
 class TableFunction : public SimpleNamedParameterFunction { // NOLINT: work-around bug in clang-tidy
 public:
+	// Overloads taking table_function_t
 	DUCKDB_API
-	TableFunction(string name, vector<LogicalType> arguments, table_function_t function,
+	TableFunction(string name, const vector<LogicalType> &arguments, table_function_t function,
 	              table_function_bind_t bind = nullptr, table_function_init_global_t init_global = nullptr,
 	              table_function_init_local_t init_local = nullptr);
 	DUCKDB_API
 	TableFunction(const vector<LogicalType> &arguments, table_function_t function, table_function_bind_t bind = nullptr,
 	              table_function_init_global_t init_global = nullptr, table_function_init_local_t init_local = nullptr);
+	// Overloads taking std::nullptr
+	DUCKDB_API
+	TableFunction(string name, const vector<LogicalType> &arguments, std::nullptr_t function,
+	              table_function_bind_t bind = nullptr, table_function_init_global_t init_global = nullptr,
+	              table_function_init_local_t init_local = nullptr);
+	DUCKDB_API
+	TableFunction(const vector<LogicalType> &arguments, std::nullptr_t function, table_function_bind_t bind = nullptr,
+	              table_function_init_global_t init_global = nullptr, table_function_init_local_t init_local = nullptr);
+
+	// Overloads taking async_table_function_t
+	DUCKDB_API
+	TableFunction(string name, const vector<LogicalType> &arguments, async_table_function_t function,
+	              table_function_bind_t bind = nullptr, table_function_init_global_t init_global = nullptr,
+	              table_function_init_local_t init_local = nullptr);
+	DUCKDB_API
+	TableFunction(const vector<LogicalType> &arguments, async_table_function_t function,
+	              table_function_bind_t bind = nullptr, table_function_init_global_t init_global = nullptr,
+	              table_function_init_local_t init_local = nullptr);
 	DUCKDB_API TableFunction();
 
+	static AsyncResultType SimpleScanImpl(table_function_t function, async_table_function_t async_function,
+	                                      ClientContext &context, TableFunctionInput &data, DataChunk &output) {
+		if (function) {
+			if (async_function) {
+				throw NotImplementedException(
+				    "Only one between function and async_function are expected, but both are provided");
+			}
+			function(context, data, output);
+			return output.size() ? SourceResultType::HAVE_MORE_OUTPUT : SourceResultType::FINISHED;
+		} else if (async_function) {
+			return async_function(context, data, output);
+		} else {
+			throw NotImplementedException(
+			    "Only one between function and async_function are expected, but none is provided");
+		}
+	}
+	bool HasSimpleScan() const {
+		return function || async_function;
+	}
+	AsyncResultType SimpleScan(ClientContext &context, TableFunctionInput &data, DataChunk &output) const {
```

Comment:
Maybe instead of adding this function, we could instead always guarantee that table functions have an `async_function`?

When registering, we could check if `async_function` is `nullptr`, then wrap `function` and place it in `async_function`. Then we could always just use `async_function` instead of `function`.